src/to-be-labelled.js

```diff
@@ -0,0 +1,124 @@
+import {matcherHint, printReceived} from 'jest-matcher-utils'
+import {checkHtmlElement, getTag} from './utils'
+
+/*
+ * Lists all tags accepting alt as an attribute
+ * @link https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
+ */
+const TAGS_WITH_ALT = ['applet', 'area', 'img']
+
+const TAGS_WITH_TITLE = ['svg']
+
+const TAGS_WITH_TITLE_ATTR = ['input', 'select', 'textarea', 'abbr']
+
+const TAGS_IMAGES = ['img', 'svg']
+
+const TAGS_WITH_TEXT_CONTENT = ['a', 'button', 'object']
+
+/*
+ * Checks if the element is labelled by `aria-label`
+ * @link https://www.w3.org/TR/WCAG20-TECHS/ARIA6.html
+ */
+function isHavingARIALabel(element) {
+  return element.hasAttribute('aria-label')
+}
+
+/*
+ * Checks if the element is labelled by `aria-labelledby`
+ * @link https://www.w3.org/TR/WCAG20-TECHS/ARIA10.html
+ */
+function isHavingARIALabelledBy(element) {
+  return element.hasAttribute('aria-labelledby')
```

Comment:
I though about that a bit but I do not know if we should do that all the time.
I am thinking about my own usage for example where I am doing a series of check on a UI library's components. This UI library is using the atomic design philosophy and for example, I want to check that my atoms can be labelled, even if the label is not existing in the tested DOM. 
In the case of my atoms & molecules components, it could complexify the test scenarii.

But at the same time, `aria-labelledby` is only a valid way to label an element if the target is existing. So I would agree with you that the test cannot really returns `true` if we are not sure the target is existing.

Should this check be an option?

Reply:
I feel that this should indeed do the additional check. If users merely wanted to check for the presence of the `aria-labelledby` attribute, they can easily do so with `.toHaveAttribute('aria-labelledby')`. The real value proposition of this matcher is to actually check the label actually exists. And as I mentioned in another comment, to eventually even allow to make assertions on the label text as well. It aligns much better with the overall value proposition of the testing libraries ecosystem to promote testing the end result that users see.

If a UI library that you're testing has components that receive the `aria-labelledby` but they do not render the target element themselves, you can always in you tests render an outer div that includes the target element with a meaningful label text, alongside your component. Then when you test it, it will actually have a label. For example in react it would be something like this:

```jsx
render(
  <div>
    <h1 id="my-component-title">Hello world</h1>
    <MyComponent aria-labelledby="my-component-title" />
  </div>
)
```

Reply:
> The real value proposition of this matcher is to actually check the label actually exists

This would be really useful - you can add this before doing 'label' queries so that you can tell what is going wrong - the label or the reference. Also ensures the aria relationships don't get stale or broken over time.